src/components/Firestore/DocumentPreview/api.tsx

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { firestore } from 'firebase';
+
+import { FirestoreArray, FirestoreAny, FirestoreMap } from '../models';
+import { getLeafKey, isMap, isArray } from '../utils';
+
+export function deleteField(
+  documentRef: firestore.DocumentReference,
+  documentData: FirestoreMap,
+  path: string[]
+) {
+  const payload = adjustPayloadForArray(
+    documentData,
+    path,
+    firestore.FieldValue.delete()
+  );
+  documentRef.update(...payload);
+}
+
+export function updateField(
+  documentRef: firestore.DocumentReference,
+  documentData: FirestoreMap,
+  path: string[],
+  value: FirestoreAny
+) {
+  const payload = adjustPayloadForArray(documentData, path, value);
+  documentRef.update(...payload);
+}
+
+/**
+ * Firestore does not allow array syntax when updating/deleting fields i.e.
+ * .update('obj.array[2]', true)
+ *
+ * This normalizes a payload's `objectPath` and `updateValue` into a form
+ * that writes to the top most array with a reconstructed payload contains the
+ * updated (or deleted) value.
+ */
+function adjustPayloadForArray(
```

Comment:
This old friend lives on! I recall console having something very similar (is this the same code?)

Reply:
Yeah with some minor tweaks, but yeah same code... probably worth a second look in a follow-up